### ISSUE TITLE
Thread.Abort is not supported in .NET Core

### ISSUE THREAD

**Issue by jadarnel27** [ErikEJ/SqlQueryStress#89]
Since the upgrade to .NET Core, the application occasionally closes without warning when I click the "Cancel" button while many threads are running.  No error message is displayed, so I pulled the source and reproduced locally while running under the Visual Studio debugger, and got this `PlatformNotSupportedException`:

> Thread abort is not supported on this platform.

At this call stack:

```
   at System.Threading.Thread.Abort()
   at SQLQueryStress.LoadEngine.ThreadKiller.KillEm() in C:\Code\SqlQueryStress\src\SQLQueryStress\LoadEngine.cs:line 610
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```

The problem is at [this line of code][1], and the reason is [documented here][2]:

> This method is obsolete. On .NET 5.0 and later versions, calling this method produces a compile-time warning. This method throws a PlatformNotSupportedException at run time on .NET 5.0 and later and .NET Core.

I'd be interested in contributing a fix, but wanted to open the issue before starting, to discuss possible solutions (don't want to overstep my bounds!).

The docs on [How to Stop a Thread][3] indicate that `CancellationToken`s should be used to cooperatively cancel threads.

However, reading through the source code, I wonder if the `Thread.Abort` could just be removed altogether.  It looks like you have cooperative cancellation built in, and this `Thread.Abort` is kind of a "last resort" if threads aren't ending quickly enough.

Removing `Thread.Abort` fixes the problem in my case - all of the threads eventually successfully cancel, and control returns to the UI

Like I said, I'm glad to contribute a fix - but wanted to make sure there's not some context I'm missing, as I'm very new to the codebase 🙂

---

My repro scenario involve spinning up 200 threads of a query that will all be blocked indefinitely.  Once they have all successfully started (I can see all 200 connections on the SQL Server side), I click cancel.  If you'd all like the details, let me know!

---

By the way, thanks for maintaining this excellent tool - it has been a huge help to me as I learn things about SQL Server.

[1]: https://github.com/ErikEJ/SqlQueryStress/blob/8d18c6dc59060bdad61e7a8482fdea68a89febf3/src/SQLQueryStress/LoadEngine.cs#L610
[2]: https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.abort?view=netcore-3.1
[3]: https://docs.microsoft.com/en-us/dotnet/standard/threading/using-threads-and-threading#how-to-stop-a-thread

**Comment by ErikEJ** [comment]
Interesting - So in other words, the "ThreadKiller should really only be concerned with closing Connections...

**Comment by jadarnel27** [comment]
@ErikEJ and calling `Cancel()` on the running `SqlCommand` objects, yeah!  "Killing" the threads should really be about cleanly ending the work the threads are doing.  Maybe the class could be renamed to `ThreadEnder` or `WorkStopper` (kidding! 😜).

As a related but separate thought, I had the idea that it might be useful to see the number of active threads in the UI.  This would give users an indication that progress is being made on the cancel operation.  I could open a separate issue for that if you think it has merit.

**Comment by ErikEJ** [comment]
Good idea, please create a separate issue for that.

**Comment by ErikEJ** [comment]
I have created a PR here: https://github.com/ErikEJ/SqlQueryStress/pull/90

**Comment by jadarnel27** [comment]
@ErikEJ I just tested that PR with my scenario, and this change has uncovered a race condition. There are two different methods calling `QueryOutInfo.CompleteAdding()` - the main `StartLoad(BackgroundWorker)` method, and the `StartLoadThread` method that represents the individual running threads.

As the last of the threads are canceling, the `StartLoad` one completes adding for that `BlockingCollection`, and then the last one or two load threads try to add items to the collection and throw errors.

I need to look into a little more closely (maybe tomorrow) to make a suggestion about the fix, but wanted to let you know there was a problem with the PR in the meantime.

**Comment by ErikEJ** [comment]
Is that not #88 you discovered?

**Comment by jadarnel27** [comment]
@ErikEJ No, I believe we introduced this here now that I've had a chance to take a closer look.  I was off about `QueryOutInfo.CompleteAdding()` being the problem, that was a symptom.

After cancelling a `SqlCommand` and closing the `SqlConnection`, the `KillEm` method *used to* loop until the `Thread` was finally dead (checking `Thread.IsAlive`).  The calling code is expecting that when the threads in `cancellerThreads` complete, that the corresponding load threads are also complete.

The simplest solution might be to add that looping behavior back in to `CancelEm`, just leaving out the `Thread.Abort` line.  This would comply with the expectations of the calling code.  I tried that locally and it worked for me.
